### PR TITLE
feature (webapi): admin api (part 7)

### LINKED ISSUE #750: LiteralV2 / StringLiteralV2 / BooleanLiteralV2
It seems that LiteralV2, StringLiteralV2, and BooleanLiteralV2 are defined in:

- `org.knora.webapi.messages.store.triplestoremessages.TriplestoreMessages.scala`
- `org.knora.webapi.messages.v2.responder.KnoraResponseV2.scala`

This makes it highly confusing to use with IntelliJ. I will rename the ones in `TriplestoreMessages` to something like `LiteralTS`, etc.

### PR DESCRIPTION
**Pull request by subotic**
### Summary

- [x] Create new list
- [x] Update list info
- [x] Check for project admin permission on create
- [x] Check for project admin permission on update
- [x] Test for required permission on create and update
- [x] Test for bad JSON payload in create and update
- [x] Rename variable

### Issues

- resolves #750

### CODE REVIEW

**Comment by subotic** [comment]
@benjamingeer this is now ready for review. thanks.

File: webapi/src/main/scala/org/knora/webapi/messages/admin/responder/listsmessages/ListsMessagesADM.scala [hunk 1]
```diff
@@ -28,6 +30,35 @@ import spray.json.{DefaultJsonProtocol, JsArray, JsObject, JsValue, JsonFormat,
...
+  *
+  * @param listIri    the IRI of the list to change.
+  * @param labels     the labels.
+  * @param comments   the comments.
+  */
+case class ChangeListInfoApiRequestADM(listIri: IRI,
+                                       labels: Option[Seq[StringLiteralV2]] = None,
+                                       comments: Option[Seq[StringLiteralV2]] = None) extends ListADMJsonProtocol {
+
+    val parametersCount: Int = List(
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/messages/admin/responder/listsmessages/ListsMessagesADM.scala]: Nice!

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala [hunk 2]
```diff
@@ -417,4 +455,159 @@ class ListsResponderADM extends Responder {
...
+      * Creates a list.
+      *
+      * @param createListRequest the new list's information.
+      * @param requestingUser the user that is making the request.
+      * @param apiRequestID   the unique api request ID.
+      * @return a [[ListInfoGetResponseADM]]
+      * @throws ForbiddenException in the case that the user is not allowed to perform the operation.
+      * @throws BadRequestException in the case when the project IRI or label is missing or invalid.
+      */
+    private def listCreateRequestADM(createListRequest: CreateListApiRequestADM, requestingUser: UserADM, apiRequestID: UUID): Future[ListGetResponseADM] = {
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala]: Does this check permissions? (Who has permission to create a list?)

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala [hunk 3]
```diff
@@ -417,4 +455,159 @@ class ListsResponderADM extends Responder {
...
+            // run list creation with an global IRI lock
+            taskResult <- IriLocker.runWithIriLock(
+                apiRequestID,
+                LISTS_GLOBAL_LOCK_IRI,
+                () => listCreateTask(createListRequest, requestingUser, apiRequestID)
+            )
+        } yield taskResult
+    }
+
+    private def listInfoChangeRequest(listIri: IRI, changeListRequest: ChangeListInfoApiRequestADM, requestingUser: UserADM, apiRequestID: UUID): Future[ListInfoGetResponseADM] = {
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala]: Who has permission to change a list?

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala [hunk 4]
```diff
@@ -417,4 +455,159 @@ class ListsResponderADM extends Responder {
...
+                dataNamedGraph = dataNamedGraph,
+                triplestore = settings.triplestoreType,
+                listIri = listIri,
+                projectIri = project.id,
+                listClassIri = OntologyConstants.KnoraBase.ListNode,
+                maybeLabels = changeListRequest.labels,
+                maybeComments = changeListRequest.comments
+            ).toString
+            // _ = log.debug("listCreateRequestADM - createNewListSparqlString: {}", createNewListSparqlString)
+            createResourceResponse <- (storeManager ? SparqlUpdateRequest(changeListInfoSparqlString)).mapTo[SparqlUpdateResponse]
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala]: Rename this variable (`createResourceResponse`)?

File: webapi/src/main/scala/org/knora/webapi/routing/admin/ListsRouteADM.scala [hunk 5]
```diff
@@ -66,7 +68,24 @@ object ListsRouteADM extends Authenticator {
             } ~
             post {
                 /* create a list */
-                ???
+                entity(as[CreateListApiRequestADM]) { apiRequest =>
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/routing/admin/ListsRouteADM.scala]: If the request is syntactically invalid (the JSON can't be parsed, or it's parsed but contains the wrong information), what sort of error does the client get?

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala (hunk 2 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala]: Good catch. No, forgot to implement this. I would say that the user needs to be a project admin.

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala (hunk 3 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/responders/admin/ListsResponderADM.scala]: dito.

File: webapi/src/main/scala/org/knora/webapi/routing/admin/ListsRouteADM.scala (hunk 5 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/routing/admin/ListsRouteADM.scala]: No idea. I will check and write a test.

**Comment by subotic** [comment]
@benjamingeer this is again ready for review. thanks.

File: webapi/src/main/scala/org/knora/webapi/routing/admin/ListsRouteADM.scala (hunk 5 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/routing/admin/ListsRouteADM.scala]: Added payload testing inside `CreateListApiRequestADM` and `ChangeListInfoApiRequestADM `. There are now also additional tests in `ListsMessagesADMSpec`

**Comment by benjamingeer** [comment]
Compile error on Travis.

File: webapi/src/main/scala/org/knora/webapi/messages/admin/responder/listsmessages/ListsMessagesADM.scala [hunk 6]
```diff
@@ -24,37 +24,77 @@ import org.knora.webapi._
...
 
     def toJsValue: JsValue = createListApiRequestADMFormat.write(this)
+
+    if (projectIri.isEmpty) {
+        // println(this)
+        throw BadRequestException(PROJECT_IRI_MISSING_ERROR)
+    }
+
+    // TODO: make check more strict, by checking if IRI is in the form `http://rdfh.ch/projects/<shortcode>`
+    // TODO: cannot use StringFormatter for this, because in ListsMessagesADMSpec we don't have a running actor system and thus no settings.
```
[the remark this answers is not in the document]
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/messages/admin/responder/listsmessages/ListsMessagesADM.scala]: cool, thanks

**Comment by subotic** [comment]
I moved MutableTestIri to the `test` folder but forgot to also add it to the `it` folder. Should be fixed with the last commit.

**Comment by subotic** [comment]
I have refactored the IRI checking to use the `StringFormatter`. This can probably be implemented better, but for now, it works.

**Comment by benjamingeer** [comment]
OK, please merge it! My grandchildren will reimplement it using quantum computing anyway. :)

**Comment by subotic** [comment]
thanks :-)

**Pull request merged**